src/analysis/schwartz_space.lean

```diff
@@ -0,0 +1,383 @@
+(f + g).seminorm_aux_le_bound k n
+  (add_nonneg (seminorm_aux_nonneg _ _ _) (seminorm_aux_nonneg _ _ _)) $
+  λ x, (decay_add_le_aux k n f g x).trans $
+  add_le_add (f.le_seminorm_aux k n x) (g.le_seminorm_aux k n x)
+
+end add
+
+section sub
+
+instance : has_sub 𝓢(E, F) :=
+⟨λ f g, ⟨f - g, (f.smooth _).sub (g.smooth _),
+  begin
+    intros k n,
+    refine ⟨f.seminorm_aux k n + g.seminorm_aux k n, λ x, _⟩,
+    refine le_trans _ (add_le_add (f.le_seminorm_aux k n x) (g.le_seminorm_aux k n x)),
+    rw sub_eq_add_neg,
+    rw ←decay_neg_aux k n g x,
+    convert decay_add_le_aux k n f (-g) x,
+    -- exact fails with deterministic timeout
+  end⟩ ⟩
+
+@[simp] lemma sub_apply {f g : 𝓢(E, F)} {x : E} : (f - g) x = f x - g x := rfl
+
+end sub
+
+section add_comm_group
+
+instance : add_comm_group 𝓢(E, F) :=
+fun_like.coe_injective.add_comm_group _ rfl (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl) (λ _ _, rfl)
+  (λ _ _, rfl)
+
+variables (E F)
+
+/-- Coercion as an additive homomorphism. -/
+def coe_hom : 𝓢(E, F) →+ (E → F) :=
+{ to_fun := λ f, f, map_zero' := coe_zero, map_add' := λ _ _, rfl }
+
+variables {E F}
+
+lemma coe_coe_hom : (coe_hom E F : 𝓢(E, F) → (E → F)) = coe_fn := rfl
+
+lemma coe_hom_injective : function.injective (coe_hom E F) :=
+by { rw coe_coe_hom, exact fun_like.coe_injective }
+
+end add_comm_group
+
+section module
+
+variables [normed_field 𝕜] [normed_space 𝕜 F] [smul_comm_class ℝ 𝕜 F]
+
+instance : module 𝕜 𝓢(E, F) :=
+coe_hom_injective.module 𝕜 (coe_hom E F) (λ _ _, rfl)
+
+end module
+
+section seminorms
+
+/-! ### Seminorms on Schwartz space-/
+
+variables [normed_field 𝕜] [normed_space 𝕜 F] [smul_comm_class ℝ 𝕜 F]
+variable (𝕜)
+
+/-- The seminorms of the Schwartz space given by the best constants in the definition of
+`𝓢(E, F)`. -/
+@[protected]
+def seminorm (k n : ℕ) : seminorm 𝕜 𝓢(E, F) := seminorm.of_smul_le (seminorm_aux k n)
+  (seminorm_aux_zero k n) (seminorm_aux_add_le k n) (seminorm_aux_smul_le k n)
+
+/-- If one controls the seminorm for every `x`, then one controls the seminorm. -/
+lemma seminorm_le_bound (k n : ℕ) (f : 𝓢(E, F)) {M : ℝ} (hMp: 0 ≤ M)
+  (hM : ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ M) : seminorm 𝕜 k n f ≤ M :=
+f.seminorm_aux_le_bound k n hMp hM
+
+/-- The seminorm controls the Schwartz estimate for any fixed `x`. -/
+lemma le_seminorm (k n : ℕ) (f : 𝓢(E, F)) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n f x∥ ≤ seminorm 𝕜 k n f :=
+f.le_seminorm_aux k n x
+
+end seminorms
+
+section topology
+
+/-! ### The topology on the Schwartz space-/
+
+variables [normed_field 𝕜] [normed_space 𝕜 F] [smul_comm_class ℝ 𝕜 F]
+variables (𝕜 E F)
+
+/-- The family of Schwartz seminorms. -/
+def _root_.schwartz_seminorm_family : seminorm_family 𝕜 𝓢(E, F) (ℕ × ℕ) :=
+λ n, seminorm 𝕜 n.1 n.2
+
+instance : topological_space 𝓢(E, F) :=
+(schwartz_seminorm_family ℝ E F).module_filter_basis.topology'
+
+lemma _root_.schwartz_with_seminorms : with_seminorms (schwartz_seminorm_family 𝕜 E F) :=
```

Comment:
Would it be significantly harder to define the topology independently of the seminorms, as an infimum of pullbacks of the topology on `bounded_continuous_function` by the linear maps I mentioned above. I used this characterization a lot when working with regular distributions so it feels wrong not to have it (although I could probably have used seminorms at the time too).